adsws/accounts/views.py

```diff
@@ -203,6 +205,119 @@ def post(self):
+                # BBB and one for API requests), when client id is not stored
+                # in the session (typically for authenticated users) we pick
+                # just the first in the database that corresponds to BBB since
+                # sessions are used by BBB and not API requests
+                client = OAuthClient.query.filter_by(user_id=session_data['user_id'], name=u'BB client').first()
+                if client:
+                    token = OAuthToken.query.filter_by(client_id=client.client_id, user_id=session_data['user_id']).first()
+                    if token:
+                        return self._translate(token.user_id, token.client_id, token.user.email, source="session:user_id")
+                    else:
+                        # If token not found in database, use data stored in session
+                        return self._translate(client.user_id, client.client_id, None, source="session:user_id")
+                else:
+                    # If client not found in database, use data stored in session
+                    return self._translate(session_data['user_id'], None, None, source="session:user_id")
+            else:
+                # This should not happen, all ADS created session should contain that parameter
+                return {'message': 'Missing oauth_client/user_id parameter in session'}, 500
+        # 2) Try to treat input data as user id
+        try:
+            user_id = int(account_data)
+        except ValueError:
+            # Try next identifier type
+            pass
+        else:
+            token = OAuthToken.query.filter_by(user_id=user_id).first()
+            if token:
+                return self._translate(token.user_id, token.client_id, token.user.email, source="user_id")
+        # 3) Try to treat input data as access token
+        token = OAuthToken.query.filter_by(access_token=account_data).first()
+        if token:
+            return self._translate(token.user_id, token.client_id, token.user.email, source="access_token")
+        # 4) Try to treat input data as client id
+        token = OAuthToken.query.filter_by(client_id=account_data).first()
+        if token:
+            return self._translate(token.user_id, token.client_id, token.user.email, source="client_id")
+        # Data not decoded sucessfully/Identifier not found
+        return {'message': 'Identifier not found'}, 404
+
+
+    def _translate(self, user_id, client_id, user_email, source=None):
+        if user_email == current_app.config['BOOTSTRAP_USER_EMAIL']:
+            anonymous = True
+        elif user_email:
+            anonymous = False
+        else:
+            anonymous = None
+
+        # 100,000 rounds of SHA-256 hash digest algorithm for HMAC (pseudorandom function)
+        # with a length of 2x32
+        hashed_user_id = binascii.hexlify(hashlib.pbkdf2_hmac('sha256', str(user_id), current_app.secret_key, 100000, dklen=32)) if user_id else None
```

Comment:
I followed the [recommendations](https://docs.python.org/2/library/hashlib.html#key-derivation). Regarding speed, link-gateway implements a caching mechanism but since session content can cache, I'm not sure how effective will be. What do you think if I implement a cache also in adsws using redis? Although, then we will have a service that can leak hashed ids to real ids relations.